build/package/scripts/cache-build.sh

```diff
@@ -0,0 +1,81 @@
+#!/bin/bash
+set -eu
+# Copy build to cache
+
+
+# the copy commands are based on GNU cp tools
+# On a mac `brew install coreutils` gives `g` prefixed cmd line tools such as gcp
+# to use these define env variable GNU_CP=gcp before invoking this script.
+CP="${GNU_CP:-cp}"
+
+OUTPUT_DIR="docker"
+WORKING_DIR="."
+CACHE_BUILD_KEY=
+CACHE_LOCATION_USED_PATH=
+DEBUG="${DEBUG:-false}"
+
+while [[ "$#" -gt 0 ]]; do
+    case $1 in
+
+    --working-dir) WORKING_DIR="$2"; shift;;
+    --working-dir=*) WORKING_DIR="${1#*=}";;
+
+    --output-dir) OUTPUT_DIR="$2"; shift;;
+    --output-dir=*) OUTPUT_DIR="${1#*=}";;
+
+    --cache-build-key) CACHE_BUILD_KEY="$2"; shift;;
+    --cache-build-key=*) CACHE_BUILD_KEY="${1#*=}";;
+
+    --cache-location-used-path) CACHE_LOCATION_USED_PATH="$2"; shift;;
+    --cache-location-used-path=*) CACHE_LOCATION_USED_PATH="${1#*=}";;
+
+    --debug) DEBUG="$2"; shift;;
+    --debug=*) DEBUG="${1#*=}";;
+
+  *) echo "Unknown parameter passed: $1"; exit 1;;
+esac; shift; done
+
+if [ -z "${CACHE_BUILD_KEY}" ]; then
+  echo "Param --cache-build-key is required."; exit 1;
+elif [ -z "${CACHE_LOCATION_USED_PATH}" ]; then
+  echo "Param --cache-location-used-path is required."; exit 1;
+fi
+
+if [ "${DEBUG}" == "true" ]; then
+  set -x
+fi
+
+ROOT_DIR=$(pwd)
+
+if [ "${WORKING_DIR}" == "." ]; then
+  git_sha_working_dir=$(git rev-parse "HEAD:")
+else
+  git_sha_working_dir=$(git rev-parse "HEAD:$WORKING_DIR")
+fi
+cache_location_dir="$ROOT_DIR/.ods-cache/build-task/$CACHE_BUILD_KEY/$git_sha_working_dir"
+
+if [ "${WORKING_DIR}" != "." ]; then
+  cd "${WORKING_DIR}"
+fi
+
+rm -rvf "$cache_location_dir"  # should be empty as otherwise cache should be used.
+mkdir -p "$cache_location_dir"
+
+# Copying reports
+cache_of_reports_dir="$cache_location_dir/reports"
+ods_artifacts_dir="${ROOT_DIR}/.ods/artifacts"
+echo "Copying build reports to cache: $ods_artifacts_dir -> $cache_of_reports_dir"
```

Comment:
I thought a bit more about caching the artifacts, since this wasn't part of the initial scope. I do understand the need to do this, but I wonder about two things:

1. the feature feels at odds with the Nexus artifact storage. Somehow now there are two locations where artifacts are temporarily persisted. Unfortunately I think Nexus is required for artifacts to work in other tasks and umbrella pipeline run. On the other hand, using Nexus for output dir caching doesn't feel right either (following the arguments you are making in the ADR)
2. I realise the code above might cache to much. Imagine if there are tasks in the pipeline before the current tasks. Their artifacts would also be in `.ods/artifacts`, and cached here. That seems wrong - I think we should only cache the artifacts of the current task. Maybe we need a "temporary artifacts folder" for the build tasks, and then move them to the right place at the end?

Reply:
Perhaps by having the artifacts copy only copy artifacts with the proper `ARTIFACT_PREFIX` derived form the working dir would help?
For variants one may also need to add the variant-key to the prefix. On the other hand I think this would already be problematic at the moment.